diagnostics-extension/src/services/events.ts

```diff
@@ -0,0 +1,169 @@
+// found in the LICENSE file.
+
+/**
+ * @fileoverview Classes related to events
+ */
+
+import { EventCategory, EventSupportStatusInfo } from '@common/dpsl';
+import { Response } from '@common/message';
+import * as fake_events from './fake_events.data';
+import { environment } from '../environments/environment';
+import { generateEventsSuccessResponse, generateErrorResponse } from '../utils';
+
+/**
+ * Abstract class reprensenting the interface of
+ * service to handle events
+ */
+export abstract class EventsService {
+  abstract registerEventHandlers(): Promise<Response>;
+  abstract registerPort(port: chrome.runtime.Port): void;
+  abstract isEventSupported(eventType: EventCategory): Promise<EventSupportStatusInfo>;
+  abstract startCapturingEvents(eventType: EventCategory): Promise<void>;
+  abstract stopCapturingEvents(eventType: EventCategory): Promise<void>;
+}
+
+/**
+ * Implementation of EventsService
+ * @extends EventsService
+ */
+export class EventsServiceImpl extends EventsService {
+  private port!: chrome.runtime.Port;
+
+  private eventCategoryAndMethods: {type: EventCategory, func: any}[] = [
+    {
+      type: EventCategory.audio_jack,
+      func: (chrome as any).os.events.onAudioJackEvent,
+    },
+    {
+      type: EventCategory.external_display,
+      func: (chrome as any).os.events.onExternalDisplayEvent,
+    },
+    {
+      type: EventCategory.keyboard_diagnostic,
+      func: (chrome as any).os.events.onKeyboardDiagnosticEvent,
+    },
+    {
+      type: EventCategory.lid,
+      func: (chrome as any).os.events.onLidEvent,
+    },
+    {
+      type: EventCategory.sd_card,
+      func: (chrome as any).os.events.onSdCardEvent,
+    },
+    {
+      type: EventCategory.stylus_connected,
+      func: (chrome as any).os.events.onStylusConnectedEvent,
+    },
+    {
+      type: EventCategory.stylus_garage,
+      func: (chrome as any).os.events.onStylusGarageEvent,
+    },
+    {
+      type: EventCategory.stylus_touch,
+      func: (chrome as any).os.events.onStylusTouchEvent,
+    },
+    {
+      type: EventCategory.touchpad_button,
+      func: (chrome as any).os.events.onTouchpadButtonEvent,
+    },
+    {
+      type: EventCategory.touchpad_connected,
+      func: (chrome as any).os.events.onTouchpadConnectedEvent,
+    },
+    {
+      type: EventCategory.touchpad_touch,
+      func: (chrome as any).os.events.onTouchpadTouchEvent,
+    },
+    {
+      type: EventCategory.touchscreen_connected,
+      func: (chrome as any).os.events.onTouchscreenConnectedEvent,
+    },
+    {
+      type: EventCategory.touchscreen_touch,
+      func: (chrome as any).os.events.onTouchscreenTouchEvent,
+    },
+    {
+      type: EventCategory.usb,
+      func: (chrome as any).os.events.onUsbEvent,
+    },
+    {
+      type: EventCategory.power,
+      func: (chrome as any).os.events.onPowerEvent,
+    },
+  ]
+
+  private notifyPort = (type: EventCategory, event) => {
+    this.port.postMessage({
```

Comment:
Should we handle case that the port is not registered here?

Reply:
Done.